### PR TITLE
meta: unify `jsonPathExtractionQuery` tests

### PR DESCRIPTION
**Pull request by WikiRik**
## Pull Request Checklist

- [X] Have you added new tests to prevent regressions?
- [X] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [X] Did you update the typescript typings accordingly (if applicable)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [X] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

This PR unifies the tests on `jsonPathExtractionQuery`

## Todos

- [ ] Decide what to do with the jsdoc
- [ ] Rename function to `_jsonPathExtractionQuery`?
- [ ] Remove `_isJson` typings in abstract and only add for postgres?
- [ ] Throw when `_isJson` is given but not supported?

### CODE REVIEW

File: src/dialects/abstract/query-generator.js [hunk 1]
```diff
@@ -1216,11 +1216,11 @@ export class AbstractQueryGenerator {
    * @param   {string}               _column   The JSON column
    * @param   {string|Array<string>} [_path]   The path to extract (optional)
    * @param   {boolean}              [_isJson] The value is JSON use alt symbols (optional)
-   * @returns {string}                        The generated sql query
+   * @returns {Error|string}                   The generated sql query
```
Comment (ephys) [inline src/dialects/abstract/query-generator.js:1219]: It never returns an error

```suggestion
   * @returns {string}                   The generated sql query
```

**Comment by WikiRik** [comment]
If it's easier I can just revert the changes in the query-generator files btw. Unifying the tests and adding types for `jsonPathExtractionQuery` would already be enough to merge this imo.

File: test/unit/query-generator/json-path-extraction-query.test.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      mysql: 'json_unquote(json_extract(`profile`,\'$.\\"id\\"\'))',
+      postgres: `("profile"#>>'{id}')`,
+      sqlite: 'json_extract(`profile`,\'$.id\')',
+    });
+  });
+
+  it('should use default handling if isJson is not passed', () => {
+    expectsql(() => queryGenerator.jsonPathExtractionQuery('profile', 'id'), {
+      default: notSupportedError,
+      mariadb: 'json_unquote(json_extract(`profile`,\'$.id\'))',
```
Comment (ephys) [inline test/unit/query-generator/json-path-extraction-query.test.ts:33]: mariadb is inconsistent with mysql: `id` is not quoted. What happens if it includes a character such as `.`, `$`, `'`, `"`, or `,`?

Note: this comment doesn't mean you have to fix it here, it can be a TODO

File: test/unit/query-generator/json-path-extraction-query.test.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      mysql: 'json_unquote(json_extract(`profile`,\'$.\\"id\\"\'))',
+      postgres: `("profile"#>>'{id}')`,
```
Comment (ephys) [inline test/unit/query-generator/json-path-extraction-query.test.ts:35]: we have too few tests for the path here. What happens if one of the entries in the path includes characters such as `'`, `}`, `,`, or `{`?

`'{id}'` is just an array. IMO we need to drop the [custom array building we have here](https://github.com/sequelize/sequelize/blob/main/src/dialects/postgres/query-generator.js#L1038) and use the stringifier implemented in the Array DataType. Having two separate implementations is going to lead to bugs

(also doesn't need to be implemented in this PR, can be a TODO)

File: test/unit/query-generator/json-path-extraction-query.test.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      sqlite: 'json_extract(`profile`,\'$.id\')',
+    });
+  });
+
+  it('should use default handling if isJson is not passed', () => {
+    expectsql(() => queryGenerator.jsonPathExtractionQuery('profile', 'id'), {
+      default: notSupportedError,
+      mariadb: 'json_unquote(json_extract(`profile`,\'$.id\'))',
+      mysql: 'json_unquote(json_extract(`profile`,\'$.\\"id\\"\'))',
```
Comment (ephys) [inline test/unit/query-generator/json-path-extraction-query.test.ts:34]: same idea as mariadb

File: test/unit/query-generator/json-path-extraction-query.test.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      postgres: `("profile"#>>'{id}')`,
+      sqlite: 'json_extract(`profile`,\'$.id\')',
```
Comment (ephys) [inline test/unit/query-generator/json-path-extraction-query.test.ts:36]: same idea as mariadb

**Review by ephys — CHANGES_REQUESTED** [review]

File: test/unit/query-generator/json-path-extraction-query.test.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+const dialectName = getTestDialect();
+
+const notSupportedError = new Error(`JSON operations are not supported in ${dialectName}.`);
+
+describe('QueryGenerator#jsonPathExtractionQuery', () => {
+  const queryGenerator = sequelize.getQueryInterface().queryGenerator;
+
+  it('should handle isJson parameter true', () => {
+    expectsql(() => queryGenerator.jsonPathExtractionQuery('profile', 'id', true), {
```
Comment (ephys) [inline test/unit/query-generator/json-path-extraction-query.test.ts:11]: other test I just thought of that we can // TODO:

if profile starts and ends with the `` ` `` character, or the `"` character (which is perfectly valid, even if an edge case), I'm pretty sure it breaks right now

File: src/dialects/abstract/query-generator.d.ts [hunk 7]
```diff
@@ -106,6 +106,11 @@ export class AbstractQueryGenerator {
     prepend?: boolean,
   ): string;
 
+  /**
+   * @private
+   */
+  jsonPathExtractionQuery(_column: string, _path?: string | string[], _isJson?: boolean): Error | string;
```
[the remark this answers is not in the document]
Comment (WikiRik) [reply src/dialects/abstract/query-generator.d.ts:137]: I'll try out the protected

The _ prefix was already there in the abstract definition, I just unified that but maybe that's not the nicest solution.

File: src/dialects/abstract/query-generator.js (hunk 1 above)
Comment (WikiRik) [reply src/dialects/abstract/query-generator.js:1219]: Yeah, I wasn't too sure about this change when I looked at it again. I think my original thought was that if it's not implemented it throws an Error, but that's not what it returns

File: test/unit/query-generator/json-path-extraction-query.test.ts (hunk 2 above)
Comment (WikiRik) [reply test/unit/query-generator/json-path-extraction-query.test.ts:33]: Yeah, working on these tests did already show some insights but I didn't really add TODOs yet. Will add those

File: src/dialects/abstract/query-generator.d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ephys) [reply src/dialects/abstract/query-generator.d.ts:137]: Because it's unused in abstract but not in other places

[the remark this answers is not in the document]
Comment (WikiRik) [reply src/dialects/abstract/query-generator.d.ts:137]: Protected kinda works, but we get the following error; `Property 'jsonPathExtractionQuery' is protected and only accessible within class 'AbstractQueryGenerator' and its subclasses.`
Should I make a new class for the tests where I extend this one? Or is there another easier solution?

[the remark this answers is not in the document]
Comment (ephys) [reply src/dialects/abstract/query-generator.d.ts:137]: Ah right we use it in tests

Let's keep your solution then

**Comment by WikiRik** [comment]
@ephys I'll leave it up to you to resolve the conversations, but I think they should all be dealt with now

**Review by ephys — APPROVED** [review]
